README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mendr

<!-- badges: start -->

[![R-CMD-check](https://github.com/Auckland-Council-CC-Insights-Team/mendr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/Auckland-Council-CC-Insights-Team/mendr/actions/workflows/R-CMD-check.yaml)
[![test-coverage](https://github.com/Auckland-Council-CC-Insights-Team/mendr/actions/workflows/test-coverage.yaml/badge.svg)](https://github.com/Auckland-Council-CC-Insights-Team/mendr/actions/workflows/test-coverage.yaml)
<!-- badges: end -->

The goal of mendr is to simplify the task of creating a calendar of
dates for a recurring process.

## Installation

You can install the development version of mendr from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("Auckland-Council-CC-Insights-Team/mendr")
```

## Example

Create a schedule for the first three months of 2023, allowing for
non-working days at Auckland Council.

``` r
library(mendr)
# Create a calendar of Auckland Council's known non-working days
create_calendar()

# Create a data frame of tasks and offsets. Alternatively, you could
# read in an Excel file here.
 tasks <- data.frame(
   tasks = c("Send email reminder", "Extract data", "Run report"),
   offsets = c(-2, 0, 1)
   )

# Pass the data frame to `create_schedule` to create a schedule for the first
# three months of 2023
create_schedule(
  schedule_data = tasks,
  starting_month = "2023-01",
  iterations = 3
  )
#>         date
#> 1 2022-12-22
#> 2 2023-01-09
#> 3 2023-01-10
#> 4 2023-01-27
#> 5 2023-02-01
#> 6 2023-02-02
#> 7 2023-02-27
#> 8 2023-03-01
#> 9 2023-03-02
```


NEWS.md
# mendr 0.1.0

* Created a basic functioning package for initial user testing
* The package has two exported functions: `create_calendar` and `create_schedule`
* Initial use-case is for creating data for use in a Power Automate flow that
sets up a schedule in Microsoft Planner


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(mendr)

test_check("mendr")


tests/testthat/test-mendr.R
test_that("we can retrieve a list of Auckland Council's non-working days", {
  expect_type(auckland_council_holidays(), "double")
  expect_equal(auckland_council_holidays() |> min(), as.Date("2022-10-24"))
  expect_equal(auckland_council_holidays() |> max(), as.Date("2024-10-28"))
})

test_that("we can create a calendar of non-working days", {
  holidays <- c(
    as.Date("2022-12-25"),
    as.Date("2022-12-26"),
    as.Date("2023-01-02"),
    as.Date("2023-01-03"),
    as.Date("2023-02-01")
  )

  create_calendar(holidays)

  expect_equal(bizdays::bizdays("2023-01-01", "2023-01-06", "NonWorkingDays"), 2)
})

test_that("we can create a schedule of tasks for several months", {
  create_calendar()

  tasks <- data.frame(
    tasks = c("Send email reminder", "Extract data", "Run report"),
    offsets = c(-2, 0, 1)
  )

  iterations <- 3

  schedule <- create_schedule(
    schedule_data = tasks,
    starting_month = "2023-01",
    iterations = iterations
  )

  # the number of rows in the schedule should equal the number of rows in our
  # dataset, multiplied by the number of iterations
  expect_equal(nrow(schedule), nrow(tasks) * iterations)
  expect_equal(max(schedule$date), as.Date("2023-03-02"))
})

test_that("we can create a schedule of tasks for one month", {
  create_calendar()

  unassigned_tasks <- data.frame(
    tasks = c("Send email reminder", "Extract data", "Run report"),
    offsets = c(-2, 0, 1)
  )

  assigned_tasks <- data.frame(
    tasks = c("Send email reminder", "Extract data", "Run report"),
    offsets = c(-2, 0, 1),
    assigned_to = c("test@test.com", NA, NA)
  )

  anchor_month <- as.Date("2023-01-01")

  expect_equal(
    schedule_month(unassigned_tasks, anchor_month) |> dplyr::pull(date) |> max(),
    as.Date("2023-01-10")
  )

  expect_equal(
    schedule_month(assigned_tasks, anchor_month) |> dplyr::pull(date) |> max(),
    as.Date("2023-01-10")
  )
})


R/mendr.R
#' List all known standard non-working days at Auckland Council
#'
#' @description `auckland_council_holidays()` is an internal function that lists
#' the dates when kaimahi at Auckland Council are not expected to work, either
#' because of public holidays or because of an organisation-wide shutdown. These
#' dates were taken from https://www.employment.govt.nz/leave-and-holidays/public-holidays/public-holidays-and-anniversary-dates/
#'
#' @return A numeric vector containing dates, beginning on 24 October 2022.
#'
#' @noRd
auckland_council_holidays <- function() {
  labour_day <- c(ymd("2022-10-24"), ymd("2023-10-23"), ymd("2024-10-28"))
  christmas_shutdown <- c(
    seq(ymd("2022-12-26"), ymd("2023-01-06"), by = "days"),
    seq(ymd("2023-12-25"), ymd("2024-01-05"), by = "days")
  )
  auckland_anniversary <- c(ymd("2023-01-30"), ymd("2024-01-29"))
  waitangi_day <- c(ymd("2023-02-06"), ymd("2024-02-06"))
  good_friday <- c(ymd("2023-04-07"), ymd("2024-03-29"))
  easter_monday <- c(ymd("2023-04-10"), ymd("2024-04-01"))
  anzac_day <- c(ymd("2023-04-25"), ymd("2024-04-25"))
  kings_birthday <- c(ymd("2023-06-05"), ymd("2024-06-03"))
  matariki <- c(ymd("2023-07-14"), ymd("2024-06-28"))

  non_working_days <- c(
    labour_day,
    christmas_shutdown,
    auckland_anniversary,
    waitangi_day,
    good_friday,
    easter_monday,
    anzac_day,
    kings_birthday,
    matariki
  )

  return(non_working_days)
}

#' Create a calendar of non-working days
#'
#' @description `create_calendar` creates a calendar and stores it in the
#'   calendar register for use when calling `create_schedule`.
#'
#' @param holidays A vector of Dates which contains the non-working days. By
#'   default, this returns a list of non-working days for staff at Auckland
#'   Council.
#' @param weekend A character vector of days that are weekends. It accepts:
#'   `sunday`, `monday`, `tuesday`, `wednesday`, `thursday`, `friday`,
#'   `saturday.` By default, Saturday and Sunday are assumed to be the days of
#'   the weekend.
#'
#' @export
#'
#' @examples
#' # A calendar of non-working days for the first Quarter of 2023, with weekends
#' # set to the default (Saturdays and Sundays)
#' q1_2023 <- c(
#'   seq(as.Date("2023-01-02"), as.Date("2023-01-06"), by = "days"),
#'   as.Date("2023-01-30"),
#'   as.Date("2023-02-06")
#' )
#' create_calendar(holidays = q1_2023)
create_calendar <- function(holidays = auckland_council_holidays(), weekend = c("sunday", "saturday")) {
  bizdays::create.calendar(
    name = "NonWorkingDays",
    holidays = holidays,
    weekdays = weekend
  )
}

#' Create a schedule of tasks
#'
#' @description `create_schedule` accepts a data frame with at least two columns
#'   (tasks and offsets), and replaces the offsets with dates.
#'
#'   # Input data
#'
#'   The first column of your data frame should contain a list of tasks, for
#'   example "Send email reminder" or "Export data". The column doesn't have to
#'   be named `task`, but it does need to be the first column in the data frame.
#'
#'   The second column of your data frame should contain a list of integers,
#'   which might be positive or negative, that define the number of working days
#'   relative to the first working day of the month. For example, 0 is the first
#'   working day of the month, 1 is the second working day of the month and so
#'   on. Alternatively, -1 is the day before the first working day of the month,
#'   -2 is two days before the first working day of the month etc. The column
#'   doesn't have to be named `offset`, but it does need to be the second column
#'   in the data frame.
#'
#'   The third column in your data frame, which is optional, should contain a
#'   list of email addresses of individuals to whom the tasks are assigned. If a
#'   task is assigned to more than one individual, you should separate their
#'   email addresses with a semi-colon (`;`). The column doesn't have to be
#'   named `assigned_to`, but it does need to be the third column in the data
#'   frame should you choose to include it.
#'
#' @param schedule_data A data frame with at least two columns, the first of
#'   type character (for tasks) and the second of type integer (for offsets). If
#'   a third column is present, it should be of type character (for email
#'   addresses).
#' @param starting_month A character string of the format `yyyy-mm` providing
#'   the starting month for the schedule.
#' @param iterations An integer specifying the number of months you wish to
#'   schedule. The default is 1.
#' @param filepath A file name to write to
#'
#' @return A [tibble()] with three columns and length equal to the number of
#'   rows in your data frame multiplied by `iterations`.
#' @export
#'
#' @examples
#' # Create a calendar of Auckland Council's known non-working days
#' create_calendar()
#'
#' # Create a data frame of tasks and offsets
#' tasks <- data.frame(
#'   tasks = c("Send email reminder", "Extract data", "Run report"),
#'   offsets = c(-2, 0, 1)
#' )
#'
#' # Pass the data frame to `create_schedule` to create a schedule for the first
#' # three months of 2023
#' create_schedule(
#'   schedule_data = tasks,
#'   starting_month = "2023-01",
#'   iterations = 3
#' )
create_schedule <- function(schedule_data, starting_month, iterations = 1, filepath = NULL) {
  first_month <- ymd(paste0(starting_month, "-01"))
  final_month <- first_month |> lubridate::add_with_rollback(months(iterations - 1))
  starting_months <- seq(first_month, final_month, by = "months")

  schedule <- purrr::map_dfr(
    .x = starting_months,
    .f = ~ schedule_month(schedule_data, anchor_month = .x)
  )

  if (!is.null(filepath)) {
    write_xlsx(schedule, path = paste(filepath, ".xlsx"))
  }

  return(schedule)
}

#' Schedule dates for a single month
#'
#' @description Given a series of tasks, offsets relative to the first working day
#'   of the month, and a starting month, convert the offsets into actual dates.
#'
#' @param schedule_data A data frame with at least two columns, the first of
#'   type character (for tasks) and the second of type integer (for offsets). If
#'   a third column is present, it should be of type character (for email
#'   addresses).
#' @param anchor_month A date, where the day is the first calendar day of the month.
#'
#' @return A [tibble()] with three columns and length equal to the number of
#'   rows (i.e. tasks) passed via `schedule_data`.
#'
#' @noRd
schedule_month <- function(schedule_data, anchor_month) {
  anchor_date <- bizdays::getdate("first bizday", anchor_month - lubridate::days(1), "NonWorkingDays")

  processed_schedule <- schedule_data |>
    mutate(
      date = bizdays::add.bizdays(anchor_date, dplyr::cur_data()[[2]], "NonWorkingDays"),
      .keep = "unused"
    )

  return(processed_schedule)
}
